### PR TITLE
add unit test

### PR DESCRIPTION
**Pull request by cadmuxe**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -11,4 +11,5 @@ install:
   - go get -u github.com/golang/dep/...
   - dep ensure
 script:
-  - make
\ No newline at end of file
+  - make
+  - make test
```
Comment (jingax10) [inline .travis.yml]: nit: not sure about the last symbol, add a blank line by the end?

File: pkg/config/config.go [hunk 2]
```diff
@@ -38,36 +37,58 @@ type Set struct {
...
+	NewChain(table, chain string) error
+	ClearChain(table, chain string) error
+	DeleteChain(table, chain string) error
+	AppendUnique(table, chain string, rulespec ...string) error
+	Delete(table, chain string, rulespec ...string) error
+}
+
 type IPTablesChainSpec struct {
 	TableName, ChainName string
 	IsDefaultChain       bool // Is a System default chain, if yes, we won't delete it.
```
Comment (jingax10) [inline pkg/config/config.go]: nit: predefined chain?

File: pkg/config/config.go [hunk 3]
```diff
@@ -38,36 +37,58 @@ type Set struct {
...
+	Ipt                  IPTabler
 }
 
 type IPTablesRuleConfig struct {
 	Spec      IPTablesChainSpec
 	RuleSpecs []IPTablesRuleSpec
+	Ipt       IPTabler
```
Comment (jingax10) [inline pkg/config/config.go]: for golang style guide, for acroym, all should be capital, e.g., Html should be HTML

File: pkg/config/config_test.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	if count, _ := ipRule.count(); len(ruleList) != 1 || count != 0 {
+		t.Error("failed to delete IPRule")
+	}
+	ipRule.Ensure(true)
+	if count, _ := ipRule.count(); len(ruleList) != 2 || count != 1 {
+		t.Error("failed to delete IPRule")
+	}
+}
+
+type MockIPtable struct {
```
Comment (jingax10) [inline pkg/config/config_test.go]: nit: s/MockIPtable/MockIPTable/

File: .travis.yml (hunk 1 above)
Comment (anfernee) [reply .travis.yml]: +1, files normally should end with a newline.

File: pkg/config/config.go [hunk 5]
```diff
@@ -38,36 +37,58 @@ type Set struct {
 	Configs     []Config
 }
 
+type Sysctler func(name string, params ...string) (string, error)
```
Comment (anfernee) [inline pkg/config/config.go]: exposed symbols should have comments, but it could be added later.

File: pkg/config/config_test.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+	ipRule := IPRuleConfig{
+		Rule:     netlink.Rule{SuppressIfgroup: -1, SuppressPrefixlen: -1, Mark: -1, Mask: -1, Goto: -1},
+		RuleAdd:  func(rule *netlink.Rule) error { ruleList = append(ruleList, *rule); return nil },
+		RuleDel:  mockRuleDel,
+		RuleList: func(family int) ([]netlink.Rule, error) { return ruleList, nil },
+	}
+	if count, _ := ipRule.count(); count != 2 {
+		t.Errorf("IPRuleConfig.count() return wrong count.")
+	}
```
Comment (anfernee) [inline pkg/config/config_test.go]: this case covered by the case underneath?

File: pkg/config/config_test.go (hunk 4 above)
Comment (anfernee) [reply pkg/config/config_test.go]: It's more like a fake than mock. the difference is subtle tho. mock only checks if a method is called and how many times it gets called. fake mimic the behavior of the interface. and fake normally requires a test for the fake itself, before it gets plugged in test.

File: pkg/config/config.go (hunk 5 above)
Comment (cadmuxe) [reply pkg/config/config.go]: Yeah, I will add comments later.

**Pull request merged**